src/main/java/dev/jbang/UnpackUtil.java

```diff
@@ -0,0 +1,153 @@
+package dev.jbang;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Enumeration;
+import java.util.LinkedHashSet;
+import java.util.Locale;
+import java.util.Set;
+
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
+
+public class UnpackUtil {
+
+	public static void unpack(Path archive, Path outputDir) throws IOException {
+		String name = archive.toString().toLowerCase(Locale.ENGLISH);
+		if (name.endsWith(".zip")) {
+			unzip(archive, outputDir, true);
+		} else if (name.endsWith(".tar.gz") || name.endsWith(".tgz")) {
+			untargz(archive, outputDir, true);
+		}
+	}
+
+	public static void unzip(Path zip, Path outputDir, boolean stripRootFolder) throws IOException {
+		try (ZipFile zipFile = new ZipFile(zip.toFile())) {
+			Enumeration<ZipArchiveEntry> entries = zipFile.getEntries();
+			while (entries.hasMoreElements()) {
+				ZipArchiveEntry zipEntry = entries.nextElement();
+				Path entry = Paths.get(zipEntry.getName());
+				if (stripRootFolder) {
+					if (entry.getNameCount() == 1) {
+						continue;
+					}
+					entry = entry.subpath(1, entry.getNameCount() - 1);
+				}
+				entry = outputDir.resolve(entry).normalize();
+				if (!entry.startsWith(outputDir)) {
+					throw new IOException("Entry is outside of the target dir: " + zipEntry.getName());
+				}
+				if (zipEntry.isDirectory()) {
+					Files.createDirectories(entry);
+				} else {
+					if (!Files.isDirectory(entry.getParent())) {
+						Files.createDirectories(entry.getParent());
+					}
+					try (InputStream zis = zipFile.getInputStream(zipEntry)) {
+						Files.copy(zis, entry);
+					}
+					int mode = zipEntry.getUnixMode();
+					if (mode != 0) {
+						Set<PosixFilePermission> permissions = PosixFilePermissionSupport.toPosixFilePermissions(mode);
+						Files.setPosixFilePermissions(entry, permissions);
+					}
+				}
+			}
+		}
+	}
+
+	public static void untargz(Path targz, Path outputDir, boolean stripRootFolder) throws IOException {
```

Comment:
did you write these methods now or copied from somewhere ? if copied lets make sure we get link to their origins (can be just)

Reply:
It's copied from code we wrote for the Launcher (which , if I remember correctly comes partly from blog posts))